File: src/main/swift/Sources/swift/fr/jhelp/kotlinLigthTools/atomic/AtomicInt.swift


public class AtomicInt {

    private let mutex: Mutex
    private var value: Int

    public init(_ value: Int){
        self.mutex = Mutex()
        self.value = value
    }

    public func get() -> Int
    {
        var value = 0
        self.mutex.safeExecute { value = self.value }
        return value
    }

    public func set(_ value: Int)
    {
        self.mutex.safeExecute { self.value = value }
    }

    public func getAndSet(_ newValue: Int) -> Int
    {
        var value = 0

        self.mutex.safeExecute {
            value = self.value
            self.value = newValue
        }

        return value
    }

    public func compareAndSet(_ excepted: Int, _ newValue: Int) -> Bool
    {
        var done = false

        self.mutex.safeExecute {
            if self.value == excepted
            {
                self.value = newValue
                done = true
            }
        }

        return done
    }

    public func incrementAndGet() -> Int
    {
        var value = 0

        self.mutex.safeExecute {
            self.value += 1
            value = self.value
        }

        return value
    }

    public func getAndIncrement() -> Int
    {
        var value = 0

        self.mutex.safeExecute {
            value = self.value
            self.value += 1
        }

        return value
    }

    public func decrementAndGet() -> Int
    {
        var value = 0

        self.mutex.safeExecute {
            self.value -= 1
            value = self.value
        }

        return value
    }

    public func getAndDecrement() -> Int
    {
        var value = 0

        self.mutex.safeExecute {
            value = self.value
            self.value -= 1
        }

        return value
    }
}

File: src/main/kotlin/fr/jhelp/kotlinLigthTools/atomic/AtomicInt.kt
package fr.jhelp.kotlinLigthTools.atomic

import fr.jhelp.kotlinLight.Mutex

class AtomicInt
{
    private val mutex: Mutex
    private var value: Int

    constructor(value: Int)
    {
        this.mutex = Mutex()
        this.value = value
    }

    fun get(): Int
    {
        var value = 0
        this.mutex.safeExecute { value = this.value }
        return value
    }

    fun set(value: Int)
    {
        this.mutex.safeExecute { this.value = value }
    }

    fun getAndSet(newValue: Int): Int
    {
        var value = 0

        this.mutex.safeExecute {
            value = this.value
            this.value = newValue
        }

        return value
    }

    fun compareAndSet(excepted: Int, newValue: Int): Boolean
    {
        var done = false

        this.mutex.safeExecute {
            if (this.value == excepted)
            {
                this.value = newValue
                done = true
            }
        }

        return done
    }

    fun incrementAndGet(): Int
    {
        var value = 0

        this.mutex.safeExecute {
            this.value++
            value = this.value
        }

        return value
    }

    fun getAndIncrement(): Int
    {
        var value = 0

        this.mutex.safeExecute {
            value = this.value
            this.value++
        }

        return value
    }

    fun decrementAndGet(): Int
    {
        var value = 0

        this.mutex.safeExecute {
            this.value--
            value = this.value
        }

        return value
    }

    fun getAndDecrement(): Int
    {
        var value = 0

        this.mutex.safeExecute {
            value = this.value
            this.value--
        }

        return value
    }
}